assets/js/blocks/product-template/edit.tsx

```diff
@@ -100,113 +161,41 @@ const ProductTemplateEdit = ( {
+		const { getTaxonomies, getEntityRecords } = select( coreStore );
+		return {
+			blocks: getBlocks( clientId ),
+			taxonomies: taxQuery
+				? getTaxonomies( {
+						type: 'product',
+						per_page: -1,
+						context: 'view',
+				  } )
+				: [],
+			templateCategory:
 				inherit &&
 				templateSlug?.startsWith( 'category-' ) &&
 				getEntityRecords( 'taxonomy', 'category', {
 					context: 'view',
 					per_page: 1,
 					_fields: [ 'id' ],
 					slug: templateSlug.replace( 'category-', '' ),
-				} );
-			const query: Record< string, unknown > = {
-				offset: perPage ? perPage * ( page - 1 ) + offset : 0,
-				order,
-				orderby: orderBy,
-			};
-			// There is no need to build the taxQuery if we inherit.
-			if ( taxQuery && ! inherit ) {
-				// We have to build the tax query for the REST API and use as
-				// keys the taxonomies `rest_base` with the `term ids` as values.
-				const builtTaxQuery = Object.entries( taxQuery ).reduce(
-					( accumulator, [ taxonomySlug, terms ] ) => {
-						const taxonomy = taxonomies?.find(
-							( { slug } ) => slug === taxonomySlug
-						);
-						if ( taxonomy?.rest_base ) {
-							accumulator[ taxonomy?.rest_base ] = terms;
-						}
-						return accumulator;
-					},
-					{}
-				);
-				if ( !! Object.keys( builtTaxQuery ).length ) {
-					Object.assign( query, builtTaxQuery );
-				}
-			}
-			if ( perPage ) {
-				query.per_page = perPage;
-			}
-			if ( author ) {
-				query.author = author;
-			}
-			if ( search ) {
-				query.search = search;
-			}
-			if ( exclude?.length ) {
-				query.exclude = exclude;
-			}
-			if ( parents?.length ) {
-				query.parent = parents;
-			}
-			// If sticky is not set, it will return all products in the results.
-			// If sticky is set to `only`, it will limit the results to sticky products only.
-			// If it is anything else, it will exclude sticky products from results. For the record the value stored is `exclude`.
-			if ( sticky ) {
-				query.sticky = sticky === 'only';
-			}
-			// If `inherit` is truthy, adjust conditionally the query to create a better preview.
-			if ( inherit ) {
-				if ( templateCategory ) {
-					query.categories = templateCategory[ 0 ]?.id;
-				}
-			}
-			return {
-				products: getEntityRecords( 'postType', postType, {
-					...query,
-					...restQueryArgs,
 				} ),
-				blocks: getBlocks( clientId ),
-			};
-		},
-		[
-			perPage,
-			page,
-			offset,
-			order,
-			orderBy,
-			clientId,
-			author,
-			search,
-			postType,
-			exclude,
-			sticky,
-			inherit,
-			templateSlug,
-			taxQuery,
-			parents,
-			restQueryArgs,
-		]
-	);
-	const blockContexts = useMemo(
-		() =>
-			products?.map( ( product ) => ( {
-				postType: product.type,
-				postId: product.id,
-			} ) ),
-		[ products ]
-	);
+		};
+	} );
+
+	const finalQuery = buildQuery( {
+		query,
+		page,
+		taxonomies,
+		templateCategory,
+	} );
+
+	const { products, productsLoading } = useStoreProducts( finalQuery );
```

Comment:
Out of curiosity, is there any reason we want to move away from `getEntityRecords`? Because recently, I came across P2 pdToLP-Aj-p2 in which Darren mentioned that we should "Converge on using the entity data store for product data in the editor" 🙂

Reply:
> I'd prefer us to implement the entity records store (with middleware) please. This will help us move towards the convergence of product block usage across multiple views (product editor, site editor, etc). It's a good opportunity to make this change.

Thanks for your input! In that case, let me start moving toward the entity records store. I was a bit skeptical after the previous attempt, but I was redirecting the request to the Store API, while maybe we could get the data from WooCommerce API ([like in this example middleware](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/utils/product-apifetch-middleware.ts)) - is that what you have in mind or do you think to still keep using Store API underneath?

> I started [experimenting with something](https://github.com/woocommerce/woocommerce-blocks/pull/9398/files#diff-010309b23c5c9005e1c87c2ed7e91a108b3e106fc44d03c807e3cb2f6ca4434bR53-R119) that could replace the withProductDataContext HOC in the price block refinement PR that might be useful to extract out to its own file. While the new ProviderFromAPI component still implements useStoreProducts, I've already noted it as a place to derive the data from the entity data store (and it'd be a good place to optionally implement a data transformation layer perhaps).

I'll definitely take a deeper look. I think that solution like that should be prepared to avoid data fetch in case the product is already provided downstream (like it's intended to be in case of Product Collection). What I mean is in the case of All Products and Products block, inner blocks are responsible to fetch the product data which is then shared through the context (`withProductDataContext `) to avoid fetching duplication. While in Product Collection the idea was to fetch all of the product data at once and just pass it downstream.